src/tiledbsoma_ml/dataset.py

```diff
@@ -5,218 +5,619 @@
+    ``obs`` rows):
+
+    1. Partitioning (|NDArrayJoinID|):
+
+        .. NOTE: for some reason, the Sphinx mathjax plugin only renders `$` blocks if at least one `:math:` directive is also present.
 
-    The ``obs_column_names`` parameter determines the data columns that are returned in the ``obs`` DataFrame (the
-    default is a single column, containing the ``soma_joinid`` for the ``obs`` dimension).
+        a. GPU-partitioning: if this is one of :math:`N>1` GPU processes (see |get_distributed_world_rank|),
+           ``obs_joinids`` is partitioned so that the $N$ GPUs will each receive the same number of samples (meaning up
+           to $N-1$ samples may be dropped). Then, only the partition corresponding to the current GPU is kept, The
+           resulting ``obs_joinids`` is used in subsequent steps.
 
-    The ``io_batch_size`` parameter determines the number of rows read, from which mini-batches are yielded. A
-    larger value will increase total memory usage and may reduce average read time per row.
+        b. |DataLoader|-worker partitioning: if this is one of $M>1$ |DataLoader|-worker processes (see
+           |get_worker_world_rank|), ``obs_joinids`` is further split $M$ ways, and only ``obs_joinids`` corresponding
+           to the current process are kept.
 
-    Shuffling support is enabled with the ``shuffle`` parameter, and will normally be more performant than using
-    :class:`DataLoader` shuffling. The shuffling algorithm works as follows:
+    2. Shuffle-chunking (|List|\[|NDArrayJoinID|\]): if ``shuffle=True``, ``obs_joinids`` are broken into "shuffle
+       chunks". The chunks are then shuffled amongst themselves (but retain their internal order, at this stage). If
+       ``shuffle=False``, one "chunk" is emitted containing all ``obs_joinids``.
 
-      1. Rows selected by the query are subdivided into groups of size ``shuffle_chunk_size``, aka a "shuffle chunk".
-      2. A random selection of shuffle chunks is drawn and read as a single I/O buffer (of size ``io_buffer_size``).
-      3. The entire I/O buffer is shuffled.
+    3. IO-batching: shuffle-chunks are re-grouped into "IO batches" of size |io_batch_size|. If ``shuffle=True``, each
+       IO-batch is shuffled, then the corresponding ``X`` and ``obs`` rows are fetched from the underlying
+       ``Experiment``.
 
-    Put another way, we read randomly selected groups of observations from across all query results, concatenate
-    those into an I/O buffer, and shuffle the buffer before returning mini-batches. The randomness of the shuffle
-    is therefore determined by the ``io_buffer_size`` (number of rows read), and the ``shuffle_chunk_size``
-    (number of rows in each draw). Decreasing ``shuffle_chunk_size`` will increase shuffling randomness, and decrease I/O
-    performance.
+    4. GPU-batching (|Iterable|\[|Batch|\]): IO-batch tuples are re-grouped into "GPU batches" of size |batch_size|.
```

Comment:
I think PyTorch typically calls these mini-batches, not GPU-batches.  I may be wrong, but is it worth adding a clarification?  E.g.,   "GPU-baching (creating mini-batches) ..."

Reply:
I would not change `batch_size`